### PR TITLE
Add components for dropdown items

### PR DESCRIPTION
**Pull request by janfaracik**
image

**Background**

Currently to create a dropdown in Jenkins you'd do something like:

```html
<l:overflowButton text="My dropdown">
  <a class="jenkins-dropdown__item" href="my-url" rel="noopener noreferrer" target="_blank">
    <div class="jenkins-dropdown__item__icon">
      <l:icon src="symbol-external" />
    </div>
    ${%Website}
  </a>
</l:overflowButton>
```

The upside of this approach is that you're not restricted as to what you put in your dropdown, e.g. you can display an image or video in your dropdown if you so wish.

The downsides of this however:
* Inconsistencies can arise between how developers implement dropdowns
* There isn't currently an easy way of having submenus in dropdowns
* There's an inconsistency between how `model-link` dropdowns work (they're powered via JSON rather than Jelly)

My proposal would be to have a set of components, e.g. `dd:item`, `dd:heading`, `dd:separator`, to be a standardised way of implementing such dropdowns. The components themselves are just simple wrappers of `template` elements, which are converted via JS into HTML to be consumed by Tippy. This gives us the advantage that our implementations of dropdowns are standardised and that our JS API for dropdowns is consistent between Jelly dropdowns and `model-link` JSON dropdowns.

```html
<template data-dropdown-type="ITEM"
          data-dropdown-id="${attrs.id}"
          data-dropdown-icon="${icon}"
          data-dropdown-text="${attrs.text}"
          data-dropdown-href="${attrs.href}"
/>
```

To allow for custom content in dropdowns I've added a `dd:custom` component which will display its child contents in the dropdown, its content needs to be escaped unlike the above components.

As part of this change to demonstrate submenus I've updated the dashboard/build history to use a dropdown, which has a submenu for RSS feeds.
image

Let me know your thoughts 👍 

### Testing done

* Tried existing dropdowns in Jenkins, all appear to work correctly
* Dashboard dropdown works, displays icon legend and submenu as expected
* There are no dropdown usages in plugins yet so now seems a good time to cement the API.

### Proposed changelog entries

- Add components for dropdown items.  Refer to the new Design Library Dropdowns page for implementation details.

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@jenkinsci/sig-ux 

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: core/src/main/resources/lib/hudson/rssBar.properties
Comment (timja) [inline core/src/main/resources/lib/hudson/rssBar.properties:27]: given these are all being changed all the other translations need removing.

(tbh they all already look broken as the keys are wrong but they should be cleaned up)

File: core/src/main/resources/lib/layout/dropdowns/custom.jelly
Comment (timja) [inline core/src/main/resources/lib/layout/dropdowns/custom.jelly:28]: maybe point to design library or give a larger example, it's not clear to be from this description how to use it

File: core/src/main/resources/lib/layout/dropdowns/separator.jelly
Comment (timja) [inline core/src/main/resources/lib/layout/dropdowns/separator.jelly:27]: add documentation to the tag?

**Comment by mawinter69** [comment]
The overflowButton tag could also need more documentation

**Comment by mawinter69** [comment]
Another thing would be to be able to provide a url from where you can fetch json a callback that can generate the html. That way you can dynamically update the menu without a page reload.  As the dropdown js things are not exposed to plugins, I basically had to reimplement it in customizable header plugin, also creating my own webpack so I can access tippy.

File: core/src/main/resources/lib/hudson/rssBar.properties
Comment (janfaracik) [reply core/src/main/resources/lib/hudson/rssBar.properties:27]: Done

File: core/src/main/resources/lib/layout/dropdowns/separator.jelly
Comment (janfaracik) [reply core/src/main/resources/lib/layout/dropdowns/separator.jelly:31]: Done

File: core/src/main/resources/lib/layout/dropdowns/custom.jelly
Comment (janfaracik) [reply core/src/main/resources/lib/layout/dropdowns/custom.jelly:28]: Added an example, will add a page to the Design Library too.

**Comment by janfaracik** [comment]
> Another thing would be to be able to provide a url from where you can fetch json or a callback that can generate the html. That way you can dynamically update the menu without a page reload. As the dropdown js things are not exposed to plugins, I basically had to reimplement it in customizable header plugin, also creating my own webpack so I can access tippy.

Definitely a good idea. There's the existing `model-link` stuff which could probably be extended to support that functionality.

**Review by Kevin-CB — APPROVED** [review]
Looks fine security wise

**Review by timja — COMMENTED** [review]
All looks good, not tested locally, could you update the design library though please?

**Review by timja — APPROVED** [review]

**Review by NotMyFault — APPROVED** [review]
Thanks!

File: war/src/main/js/components/dropdowns/templates.js
Comment (daniel-beck) [inline war/src/main/js/components/dropdowns/templates.js:48]:
```suggestion
      <${tag} class="jenkins-dropdown__item ${itemOptions.clazz ?: ""}" href="${itemOptions.url}" id="${itemOptions.id}">
```

JEXL supports Elvis operator. Untested.

File: war/src/main/js/components/dropdowns/utils.js
Comment (daniel-beck) [inline war/src/main/js/components/dropdowns/utils.js:153]: This looks like the likely culprit of a lot of `id="undefined"`?

File: war/src/main/resources/images/symbols/information-circle.svg
Comment (daniel-beck) [inline war/src/main/resources/images/symbols/information-circle.svg:1]: Why is this change part of this PR, and why is it undocumented? AFAICT this icon is used in the sidepanel of agents mostly. It also looks wrong, the "i" seems off center.

> Screenshot 2024-02-19 at 19 36 46

> Screenshot 2024-02-19 at 19 36 48

**Review by daniel-beck — COMMENTED** [review]
The custom tags look a lot better than the HTML, and a new taglib seems like a good location 👍 

Re no changelog, this should have at least a _Developer:_ entry.

**Review by daniel-beck — DISMISSED** [review]
See previous review

File: war/src/main/js/components/dropdowns/templates.js
Comment (janfaracik) [reply war/src/main/js/components/dropdowns/templates.js]: Updated it - will see if it passes unil tests

File: war/src/main/js/components/dropdowns/utils.js
Comment (janfaracik) [reply war/src/main/js/components/dropdowns/utils.js:153]: Fixed - good spot!

File: war/src/main/resources/images/symbols/information-circle.svg
Comment (janfaracik) [reply war/src/main/resources/images/symbols/information-circle.svg:1]: Reverted - thanks!

File: war/src/main/js/components/dropdowns/templates.js
Comment (janfaracik) [reply war/src/main/js/components/dropdowns/templates.js]: Unit tests weren't happy sadly, reverted it.

Comment (mawinter69) [reply war/src/main/js/components/dropdowns/templates.js]: Does javascript support the Elvis operator? This is not jelly

Comment (janfaracik) [reply war/src/main/js/components/dropdowns/templates.js]: It doesn't - I've replaced it with `||` and it looks to be working fine. Also updated it to remove `href` being added when its a button rather than a link. Thanks!

Comment (daniel-beck) [reply war/src/main/js/components/dropdowns/templates.js]: > This is not jelly

Ah, good point, it looked so similar to JEXL/Jelly that I thought it was that. Sorry for the noise!

**Comment by janfaracik** [comment]
Anybody have any other blockers with this PR or is it good to go?

File: war/src/main/js/components/dropdowns/templates.js
Comment (daniel-beck) [inline war/src/main/js/components/dropdowns/templates.js:50]: Should be `xmlEscape`'d (probably including the URL even though it's not new here).

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Comment by basil** [comment]
Caused https://github.com/jenkinsci/pipeline-graph-view-plugin/issues/391

**Pull request merged**
